KAFKA-14351: Controller Mutation Quota for KRaft

rondagostino opened the pull request:
Check controller mutation quotas in KRaft.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rondagostino committed:
KAFKA-14351: Controller Mutation Quota for KRaft

rondagostino commented on docs/ops.html:
```diff
@@ -1197,7 +1197,7 @@ <h4 class="anchor-heading"><a id="multitenancy-isolation" class="anchor-link"></
 
   <p>
-  <strong>Client quotas:</strong> Kafka supports different types of (per-user principal) client quotas. Because a client's quotas apply irrespective of which topics the client is writing to or reading from, they are a convenient and effective tool to allocate resources in a multi-tenant cluster. <a href="#design_quotascpu">Request rate quotas</a>, for example, help to limit a user's impact on bro ...
+  <strong>Client quotas:</strong> Kafka supports different types of (per-user principal) client quotas. Because a client's quotas apply irrespective of which topics the client is writing to or reading from, they are a convenient and effective tool to allocate resources in a multi-tenant cluster. <a href="#design_quotascpu">Request rate quotas</a>, for example, help to limit a user's impact on bro ...
```
Hard to see the change, which is on the last line: changes `controller_mutations_rate` to `controller_mutation_rate` since the latter is what was implemented despite what the KIP says.

cmccabe replied:
ack

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/ControllerRequestContext.java:
```diff
@@ -42,19 +45,39 @@ public static OptionalLong requestTimeoutMsToDeadlineNs(
     private final OptionalLong deadlineNs;
     private final RequestHeaderData requestHeader;
 
+    private final Optional<Consumer<Double>> requestedPartitionCountRecorder;
+
```
We add this here because `class ControllerMutationQuotaManager` and `trait ControllerMutationQuota` are written in Scala while the KRaft controller and such are written in Java with no access to the Scala class/trait -- we can't pass in anything without wrapping it in Java somehow.  All we need to do is record a request for some number of partitions, and this matches that requirement precisely.

ijuma replied:
We could have an interface in server-common that the scala classes implement.

cmccabe replied:
Yes, this is messy. I have a lot of ideas for improving this, but most of them would require substantial refactoring.

For now, let's just get rid of the obvious bad stuff.
* Using a double for number of partitions is a bit silly
* "Recorder" is also a bit of a misnomer -- we are doing more than recording the number of partitions.
* The Optional isn't needed -- just pass a Consumer that does nothing if you don't want to do anything.

So maybe change it to something like `Consumer<Integer> applyPartitionChangeQuota` ?

rondagostino commented:
Only implemented/tested the create-topics case so far.

cmccabe commented:
Thanks very much for this, @rondagostino ! 

Anyway what I’d like to do here is two things:
1. split all the spelling corrections and docs changes into a separate PR which we can do today. They look like no-brainers

2. second PR do everything in Scala, see if that is cleaner. I think it will be. It would be better not to involve the core controller in this. we should be able to see which partition(s) were successfully created from ControllerServer (check the controller response)

also we'll need a test. As you said, we need to get a bit of dynamic config code in first, but that is pretty small.

rondagostino commented:
Thanks, Colin.  I moved the minor stuff into https://github.com/apache/kafka/pull/13118 and put the check into `ControllerApis`.  The only difference with what you suggested is that I consider the partitions requested (except for the ones with obvious issues like lack of authorization or duplicate topic names) rather than the partitions that the controller creates.  When a quota is exceeded the request is not accepted, so we can't let the controller mutate the partition count and then decide -- we have to decide before letting the controller do its thing.

Again, the create-topic only case is the only one I've done.  I've added a test as well.

rondagostino commented on core/src/test/scala/unit/kafka/server/ControllerApisTest.scala:
```diff
@@ -523,8 +523,23 @@ class ControllerApisTest {
+  }
+
+  @ParameterizedTest(name = "testCreateTopics with mutationQuotaExceeded: {0}, validateOnly: {1}")
+  @CsvSource(value = Array(
+    "no,no",
+    "no,yes",
+    "yes,no",
+    "yes,yes"
+  ))
```
Instead of `yes/no` for the quota maybe this should be an int so that we can test that we are actually sending the correct partition count. We are requesting 6 potentially-valid partitions, so it might be good to ensure that a limit of 6 is throttled but a limit of 7 is not.

rondagostino commented:
@cmccabe I updated this to deal with create topic requests that either set an explicit assignment or that use the default partition count.  Note that this is logic that exists within `QuorumController`, and to take it into account with the throttling logic being invoked in Scala means we have to have it in `ControllerApis`.  Also note that I didn't implement the full logic -- `QuorumController` has additional checks that could in fact cause a specific topic request to be rejected (e.g. explicit assignments not contiguous partition numbers starting with 0).  This could lead to a calculation of partition counts for throttling purposes that exceeds the actual partition counts requested -- which could cause throttling in cases where technically it should not.  I think this would be a rare occurrence -- so a small detriment to the Scala approach.

The benefit of the Scala approach is that we avoid doing work in the controller itself when throttling does occur.  We don't generate all of the Metadata Log records, for example.  Note that createPartitions and delateTopics will have to get access to the partition counts, which currently I think is not available.

cmccabe commented on core/src/main/scala/kafka/server/ControllerApis.scala:
```diff
@@ -392,14 +394,36 @@ class ControllerApis(val requestChannel: RequestChannel,
+          data.topics.add(new CreateTopicsResponseData.CreatableTopicResult().setName(topic.name).setErrorCode(apiError.error.code).setErrorMessage(apiError.message)))
+        data.setThrottleTimeMs(e.throttleTimeMs())
+        CompletableFuture.completedFuture(data)
+    }
```
I don't think you need to catch this here, since we have a catch block on the request handler that will deal with "send back error X to response Y"

If you do have a catch block here you'd certainly need a return statement as well

rondagostino replied:
I actually believe the code is correct as written.  We mark all of the requested topics in `effectiveRequest` with the throttle exception, and then we fallthrough to mark all of the topics that were not in the effective request with the appropriate error codes (duplicate topics get marked with invalid request and unauthorized topics get marked with unauthorized request).  Clearly this is not clear though!  I'll add a comment describing this.

rondagostino commented:
Thanks, Colin.  That all makes sense.  The KIP indicates that `validateOnly` is not throttled because it does not load the controller (https://cwiki.apache.org/confluence/display/KAFKA/KIP-599%3A+Throttle+Create+Topic%2C+Create+Partition+and+Delete+Topic+Operations#KIP599:ThrottleCreateTopic,CreatePartitionandDeleteTopicOperations-HandlingofValidateOnly), but that is not the case here as you point out.  So I will proceed as follows:

1. Keep the logic inside `ControllerApis.scala` to avoid taking locks in the controller thread
2. Expose any necessary information required via the `org.apache.kafka.controller.Controller` interface.

It isn't clear to me that we should charge for unauthorized requests.  First, this is not mentioned in the KIP, so it is a different case than the `validateOnly` situation (where the spirit of the KIP implies we should charge for `validateOnly` in KRaft).  Also, I wonder if it is possible that charging for unauthorized requests could lead to indirectly exposing partition information that a client is unauthorized to know.  For now I will leave this piece alone.

cmccabe commented on core/src/main/scala/kafka/server/ControllerApis.scala:
```diff
@@ -322,15 +325,37 @@ class ControllerApis(val requestChannel: RequestChannel,
-        controller.deleteTopics(context, idToName.keySet).thenApply { idToError =>
-          idToError.forEach { (id, error) =>
-            appendResponse(idToName.get(id), id, error)
+        // First check the controller mutation quota if necessary, and then
```
This is a lot more logic (and expense) than what I wanted, and unfortunately it also has a TOCTOU, because partitions may be added in between checking here and doing the deletion.

I guess I simply didn't forsee this, sorry.

I wonder if we could convert the throttle to be lockless so that we could use it in the QuorumController thread itself. Like I said earlier, the big concern is that blocking the controller thread is really bad and not something we want to do under really any condition. It seems like the throttle should be simple... just a number or something right? Or can we somehow guarantee that only the controller thread itself is taking that lock under ordinary conditions?

rondagostino replied:
I believe these locks are scoped by the controller mutation quota manager such that nobody else would contest for them aside from the single-threaded controller.  @dajac might you be able to confirm this?

dajac replied:
The quota is basically a `Sensor` which would only be updated by the QuorumController thread. However, I suppose that the internal metric lock is also acquired when the metric is reported so the QuorumController would not be the only one trying to acquire it, I think.

rondagostino committed:
Add delete topic support

rondagostino committed:
Add create partition support

rondagostino committed:
Test throttle time set correctly in ControllerApis

rondagostino committed:
whitespace fixes

rondagostino committed:
Add dynamic reconfiguration, no tests yet

rondagostino committed:
Revert "Add dynamic reconfiguration, no tests yet"

This reverts commit 7722647804507ef48da0adfb4de956c29021bed7.

rondagostino committed:
Add dynamic reconfiguration of configs and client quotas

rondagostino committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-14351

rondagostino committed:
Register controller for dynamic KafkaConfig changes

rondagostino committed:
Add system test, fix combined mode

rondagostino commented:
```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.11.1
session_id:       2023-03-05--001
run time:         1 minute 33.639 seconds
tests run:        3
passed:           3
flaky:            0
failed:           0
ignored:          0
================================================================================
test_id:    kafkatest.tests.core.controller_mutation_quota_test.ControllerMutationQuotaTest.test_controller_mutation_quota.metadata_quorum=COLOCATED_KRAFT
status:     PASS
run time:   26.262 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.controller_mutation_quota_test.ControllerMutationQuotaTest.test_controller_mutation_quota.metadata_quorum=REMOTE_KRAFT
status:     PASS
run time:   30.448 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.controller_mutation_quota_test.ControllerMutationQuotaTest.test_controller_mutation_quota.metadata_quorum=ZK
status:     PASS
run time:   36.590 seconds
--------------------------------------------------------------------------------
```

rondagostino committed:
Add DynamicBrokerConfig test for controllers

rondagostino committed:
Fix log message

rondagostino committed:
testReconfigureControllerClientQuotas

rondagostino committed:
put the addBrokerReconfigurable() calls back in

cmccabe commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -730,6 +744,45 @@ class DynamicThreadPool(server: KafkaBroker) extends BrokerReconfigurable {
+class ControllerDynamicThreadPool(controller: ControllerServer) extends BrokerReconfigurable {
+
+  override def reconfigurableConfigs: Set[String] = {
+    DynamicThreadPool.ReconfigurableConfigs // common configs
```
This is a small thing but, we probably shouldn't return `num.replica.fetchers`, `num.recovery.threads.per.data.dir`, and `background.threads` here, since they're not actually reconfigurable on the controller.

We could still share most of the code here if we pass in the `ReconfigurableConfigs` `Set[String]` to `DynamicThreadPool.validateReconfiguration`, right?

cmccabe commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -911,7 +954,22 @@ object DynamicListenerConfig {
 }
 
-class DynamicClientQuotaCallback(server: KafkaBroker) extends Reconfigurable {
+trait DynamicClientQuotaCallbackServer {
```
We don't really need a trait for this, do we? It's just two objects that we can pass directly to the constructor of `DynamicClientQuotaCallback`

cmccabe commented on core/src/test/java/kafka/test/MockController.java:
```diff
@@ -181,10 +194,16 @@ public CompletableFuture<Void> unregisterBroker(
+        private final int numPartitions;
 
         MockTopic(String name, Uuid id) {
+            this(name, id, 10); // hard-code the number of partitions if left unspecified
```
I guess this is a nitpick but `num.partitions` defaults to 1, so having a default of 10 partitions is a bit surprising here. Is there a reason to use 10 instead of 1?

cmccabe commented on core/src/main/scala/kafka/server/metadata/DynamicClientQuotaPublisher.scala:
```diff
@@ -0,0 +1,82 @@
+  faultHandler: FaultHandler,
+  nodeType: String,
+  clientQuotaMetadataManager: ClientQuotaMetadataManager,
+) extends Logging with org.apache.kafka.image.publisher.MetadataPublisher {
```
I didn't think about this previously, but it makes sense to include `controller` or `broker` in the name that is returned here. Good idea.

cmccabe commented on core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:
```diff
@@ -479,6 +489,42 @@ class DynamicBrokerConfigTest {
+    when(controllerServer.config).thenReturn(oldConfig)
+    when(controllerServer.authorizer).thenReturn(Some(authorizer))
+    // We are only testing authorizer reconfiguration, ignore any exceptions due to incomplete mock
+    assertThrows(classOf[Throwable], () => controllerServer.config.dynamicConfig.addReconfigurables(controllerServer))
```
I sympathize with not wanting to debug the mocking mess here. There are just so many functions that get called, and having to do thenReturn for all of them is a huge pain. Still, just putting an `assertThrows(classOf[Throwable], ...)` seems like it could hide errors over time.

Should we just add the authorizer as a reconfigurable directly? Or is that too hard to do here?

rondagostino replied:
I fixed it.

cmccabe commented on core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:
```diff
@@ -479,6 +489,42 @@ class DynamicBrokerConfigTest {
+    when(controllerServer.config).thenReturn(oldConfig)
+    when(controllerServer.authorizer).thenReturn(Some(authorizer))
+    // We are only testing authorizer reconfiguration, ignore any exceptions due to incomplete mock
+    assertThrows(classOf[Throwable], () => controllerServer.config.dynamicConfig.addReconfigurables(controllerServer))
```
Same question as above

rondagostino replied:
Same here -- I fixed it.

cmccabe commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -378,6 +378,118 @@ object TestUtils extends Logging {
     props
   }
 
+  def createCombinedControllerConfig(nodeId: Int,
```
Can't we just add a boolean to `TestUtils.createBrokerConfig`? Two whole new functions seems like overkill.

Also, while you're changing this code, it would be good to use the standard indentation style. Having the parameter list be indented halfway through the page really hurts the eyes. Thsee days we do:

```
def myVeryLongFunctionName(
  foo: Foo,
  bar: Bar,
...
  quux: Quux
): ReturnValueType = {
  ...
}
```

ijuma replied:
`createBrokerConfig` has a ton of parameters, I don't think we should be adding more parameters to it.

cmccabe commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -378,6 +378,118 @@ object TestUtils extends Logging {
+    retval
+  }
+
+  def createIsolatedControllerConfig(nodeId: Int,
```
Same question as above, why not just a boolean on `createBrokerConfig`?

`combinedMode: Boolean = false` or similar

cmccabe commented:
Thanks for the PR, @rondagostino . You've been very patient here and I'm sorry that the review wasn't quicker. We did a major pivot from implementing the quota in `ControllerApis` like I originally wanted, to implementing it in `QuorumController` like you originally wanted. And I think the QC way is the way it has to be, for all the reasons discussed above. So you were right all along :)

I spent a few hours looking through this code today to see if there were any obvious ways to improve the performance characteristics. The main thing I worry about is the impact of locking. The locking in `StrictControllerMutationQuota.record`kj is certainly frustrating, since I could see it colliding with metrics collection. There are several small locks that might be contended (sensor locks, KafkaMetric locks). None of them seems to be held for a very long time, though. So overall I cannot find any easy way to avoid this worry. We will have to benchmark, I guess.

I left some minor comments that probably you can do in a day or two. I think we're pretty close now.

cmccabe committed:
fixup

rondagostino committed:
Colin's fixups -- thanks!

rondagostino commented:
There were 6 test failures across 3 separate builds.  All tests pass locally.

rondagostino committed:
Fix a couple of nits as per review

rondagostino committed:
Fix Scala 2.12 compile errors

rondagostino commented:
There were several test failures (SelectorTest, Tls12SelectorTest, KafkaServerKRaftRegistrationTest, ListOffsetsRequestWithRemoteStoreTest, SocketServerTest) but I have seen them in the recent past sporadically, and they all passed locally. Some builds still running/

cmccabe commented:
Yes. It is frustrating that we have so many flaky tests. @mumrah is fixing the `KafkaServerKRaftRegistrationTest`. We may need to fix `SocketServerTest` too in the near future since it seems to flake a lot. I agree that the test failures are unrelated. Will commit.
